### PR TITLE
VPN-6116 - Implement missing subscription events

### PR DESCRIPTION
**Pull request by brizental**

### CODE REVIEW

**Comment by brizental** [comment]
DATA REVIEW REQUEST
1. What questions will you answer with this data?

- How often do started subscription attempts fail?
- How often do started subscription attempts succeed?

2. Why does Mozilla need to answer these questions? Are there benefits for users?
   Do we need this information to address product or business requirements?

We need to understand the health of the subscription pipeline.

3. What alternative methods did you consider to answer these questions?
   Why were they not sufficient?

None.

4. Can current instrumentation answer these questions?

No.

5. List all proposed measurements and indicate the category of data collection for each
   measurement, using the Firefox data collection categories found on the Mozilla wiki.

Measurement Name | Measurement Description | Data Collection Category | Tracking Bug
---------------- | ----------------------- | ------------------------ | ------------
`outcome.subscription_already_present` | The subscription was blocked because the user is already a subscriber.  | interaction | https://mozilla-hub.atlassian.net/browse/VPN-6116
`outcome.subscription_failed` | The subscription process has errored.  | interaction | https://mozilla-hub.atlassian.net/browse/VPN-6116

6. Please provide a link to the documentation for this data collection which
   describes the ultimate data set in a public, complete, and accurate way.

This collection is Glean so is documented [in the Glean Dictionary](https://dictionary.telemetry.mozilla.org).

7. How long will this data be collected?

This collection will be collected permanently.
brizental@mozilla.com, vpn-telemetry@mozilla.com will be responsible for the permanent collections.

8. What populations will you measure?

All channels, countries, and locales. No filters.

9. If this data collection is default on, what is the opt-out mechanism for users?

These collections are Glean. The opt-out can be found in the product's preferences.

10. Please provide a general description of how you will analyze this data.

Using Looker dashboards.

11. Where do you intend to share the results of your analysis?

Withing the Mozilla VPN team and PXI org.

12. Is there a third-party tool (i.e. not Glean or Telemetry) that you
    are proposing to use for this data collection?

No.

**Comment by chutten** [comment]
DATA COLLECTION REVIEW RESPONSE:

    Is there or will there be documentation that describes the schema for the ultimate data set available publicly, complete and accurate?

Yes.

    Is there a control mechanism that allows the user to turn the data collection on and off?

Yes. This collection can be controlled through the product's preferences.

    If the request is for permanent data collection, is there someone who will monitor the data over time?

Yes, Bea and the VPN Team are responsible.

    Using the category system of data types on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Category 2, Interaction.

    Is the data collection request for default-on or default-off?

Default on for all channels.

    Does the instrumentation include the addition of any new identifiers?

No.

    Is the data collection covered by the existing Firefox privacy notice?

Yes.

    Does the data collection use a third-party collection tool?

No.

---
Result: datareview+

**Review by chutten — COMMENTED** [review]
(`data-review+`)

File: tests/functional/testSubscriptionNeeded.js [hunk 1]
```diff
@@ -105,7 +105,26 @@ describe("subscription needed tests", function() {
...
+        // Click the "Subscribe now" button
+        await vpn.waitForQueryAndClick(queries.screenSubscriptionNeeded.SUBSCRIPTION_NEEDED_BUTTON.visible());
+        const startedEventsList = await vpn.gleanTestGetValue("outcome", "subscriptionStarted", "main");
+        assert.strictEqual(startedEventsList.length, 1);
+
+        // Subscription, for Guardian, is the same as in-browser auth.
+        await vpn.mockInBrowserAuthentication();
+
+        const completedEventsList = await vpn.gleanTestGetValue("outcome", "subscriptionCompleted", "main");
+        assert.strictEqual(completedEventsList.length, 1);
```
Comment (brizental) [inline tests/functional/testSubscriptionNeeded.js:120]: This is not working and also this is missing tests for failed states.

**Review by mcleinman — CHANGES_REQUESTED** [review]
I know there have been concerns about renaming metrics before, and that it may make it hard to connect the old metric and the renamed one. Are these new metric names supposed to be an exact substitute for the old ones? If so, is data on board with this?

Also, we're not recording anything for these situations - should we be?
- subscriptionCancelled
- billingNotAvailable
- subscriptionCompleted (Should we be recording subscriptionCompleted metric in here? I don't see it used anywhere.)

**Comment by brizental** [comment]
@mcleinman This is now ready for re-review. I see your point about subscriptionCancelled and billingNotAvailable. However neither of these event is listed in the Miro board (https://miro.com/app/board/uXjVM0BZcnc=/?userEmail=sandrigo@mozilla.com&openComment=3458764559943493792&mid=8418131&utm_source=notification&utm_medium=email&utm_campaign=mentions&utm_content=reply-to-mention). This might just be an oversight, but even then I'd consider it our of scope for this PR.

File: tests/functional/testSubscriptionNeeded.js (hunk 1 above)
Comment (brizental) [reply tests/functional/testSubscriptionNeeded.js]: As commented on https://github.com/mozilla-mobile/mozilla-vpn-client/pull/9150#issuecomment-2090057667 failure events are out of scope for this ticket.

**Comment by mcleinman** [comment]
If it's an oversight, now is the best time to fix it (rather than having partial data for a while). If it's intentional, this seems great. 

I'd want an affirmative confirmation from Santiago that he's okay without those metrics before approving.

File: tests/functional/testAuthenticationInBrowser.js [hunk 2]
```diff
@@ -28,7 +28,7 @@ describe('User authentication in browser', function() {
     await vpn.clickOnQuery(
         queries.screenInitialize.AUTHENTICATE_VIEW.visible());
     await vpn.waitForQueryAndClick(
-        queries.screenAuthenticating.CANCEL_FOOTER_LINK.visible());
+        queries.screenAuthenticating.CANCELe_FOOTER_LINK.visible());
```
Comment (brizental) [inline tests/functional/testAuthenticationInBrowser.js:31]:
```suggestion
        queries.screenAuthenticating.CANCEL_FOOTER_LINK.visible());
```

**Comment by albionx** [comment]
Ticket updated here: https://mozilla-hub.atlassian.net/browse/VPN-6116

File: src/telemetry/outcome_metrics.yaml [hunk 3]
```diff
@@ -153,6 +153,39 @@ outcome:
...
+    notification_emails:
+      - vpn-telemetry@mozilla.com
+      - brizental@mozilla.com
+    expires: never
+    extra_keys:
+      reason:
+        description: |
+          What caused the subscription failure.
+          Possible values are:
+            - BillingaNotAvailable (Android-only)
```
Comment (mcleinman) [inline src/telemetry/outcome_metrics.yaml:178]:
```suggestion
            - BillingNotAvailable (Android-only)
```

File: tests/functional/testAuthenticationInBrowser.js (hunk 2 above)
Comment (mcleinman) [reply tests/functional/testAuthenticationInBrowser.js:31]: Doesn't this need to be updated still?

**Review by mcleinman — APPROVED** [review]
R+ 
Two small typos that may need updates. Thanks!

**Pull request merged**